certbot-apache/certbot_apache/interfaces.py

```diff
@@ -110,34 +112,41 @@ class ParserNode(object):
     it's responsibility of each ParserNode object itself to store its prepending
     whitespaces in order to be able to reconstruct the complete configuration file
     as it was when originally read from the disk.
-    """
 
-    @property
-    @abc.abstractmethod
-    def ancestor(self):  # pragma: no cover
-        """
-        This property contains a reference to ancestor node, or None if the node
-        is the root node of the configuration tree.
+    ParserNode objects should have the following attributes:
 
-        :returns: The ancestor BlockNode object, or None for root node.
-        :rtype: ParserNode
-        """
+    # Reference to ancestor node, or None if the node is the root node of the
+    # configuration tree.
+    ancestor: Optional[ParserNode]
 
-        raise NotImplementedError
+    # True if this node has been modified since last save.
+    dirty: bool
+
+    # Filepath of the file where the configuration element for this ParserNode
+    # object resides. For root node, the value for filepath is the httpd root
+    # configuration file. Filepath can be None if a configuration directive is
+    # defined in for example the httpd command line.
+    filepath: Optional[str]
+    """
 
-    @property
     @abc.abstractmethod
-    def dirty(self):  # pragma: no cover
+    def __init__(self, **kwargs):
         """
-        This property contains a boolean value of the information if this node has
-        been modified since last save (or after the initial parse).
+        Initializes the ParserNode instance, and sets the ParserNode specific
+        instance variables. This is not meant to be used directly, but through
+        specific classes implementing ParserNode interface.
 
-        :returns: True if this node has had changes that have not yet been written
-            to disk.
-        :rtype: bool
-        """
+        :param ancestor: BlockNode ancestor for this CommentNode. Required.
+        :type ancestor: BlockNode or None
 
-        raise NotImplementedError
+        :param filepath: Filesystem path for the file where this CommentNode
+            does or should exist in the filesystem. Required.
+        :type filepath: str or None
+
+        :param dirty: Boolean flag for denoting if this CommentNode has been
+            created or changed after the last save. Default: False.
+        :type dirty: bool or None
```

Comment:
If `dirty` is set, it should always be set to a boolean.
```suggestion
        :type dirty: bool
```